From e28a134e847eff8bc6a27ca7b0bdde424716497f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pasi=20Pet=C3=A4j=C3=A4j=C3=A4rvi?= <pasi.petajajarvi@qt.io>
Date: Tue, 30 Mar 2021 17:40:03 +0300
Subject: [PATCH] eglfs: Newer Nvidia libdrm provide device instead driver
 module name

In newer Nvidia proprietary libdrm binaries display device name is
actual device and not driver module name. Check which provided device
name has been returned with EGL_DRM_DEVICE_FILE_EXT to choose correct
function to open device.

Pick-to: 5.15
Fixes: QTBUG-91184
Change-Id: I95f907dfa30057da0dca4ff32e0605c6bb10e0a5
Reviewed-by: Laszlo Agocs <laszlo.agocs@qt.io>
---
 .../eglfs_kms_egldevice/qeglfskmsegldevice.cpp             | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_egldevice/qeglfskmsegldevice.cpp b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_egldevice/qeglfskmsegldevice.cpp
index 81ce7e1d71..749750042c 100644
--- a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_egldevice/qeglfskmsegldevice.cpp
+++ b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_egldevice/qeglfskmsegldevice.cpp
@@ -58,7 +58,12 @@ bool QEglFSKmsEglDevice::open()
 {
     Q_ASSERT(fd() == -1);
 
-    int fd = drmOpen(devicePath().toLocal8Bit().constData(), nullptr);
+    int fd = -1;
+
+    if (devicePath().compare("drm-nvdc") == 0)
+        fd = drmOpen(devicePath().toLocal8Bit().constData(), nullptr);
+    else
+        fd = qt_safe_open(devicePath().toLocal8Bit().constData(), O_RDWR);
     if (Q_UNLIKELY(fd < 0))
         qFatal("Could not open DRM (NV) device");
 
